### PR TITLE
Aligning with changes in BHoMAdapter Refactoring Level 04

### PR DESCRIPTION
**Pull request by alelom**
### NOTE: Depends on 

https://github.com/BHoM/BHoM_Engine/pull/1371   
https://github.com/BHoM/BHoM_Adapter/pull/164
   
### Issues addressed by this PR

Closes https://github.com/BHoM/BHoM_UI/issues/171
Closes https://github.com/BHoM/BHoM_UI/issues/166

### Changelog

- Added PushType/PullType to the respective method inputs
- Delete action renamed to Remove
- Adapter actions now accepts IRequest 
- General tidy up of UI methods for adapter actions

### CODE REVIEW

**Comment by adecler** [comment]
Just noting that the `BHoM_UI.ChekCore` test fails because Dynamo doesn't yet have a corresponding PR.

**Comment by epignatelli** [comment]
Is this ready to review, or should I wait?

File: BHoM_UI/Components/Adapter/Move.cs [hunk 1]
```diff
@@ -67,13 +68,30 @@ public class MoveCaller : MethodCaller
         [Input("pushConfig", "Config options for the Push to Target adapter")]
         [Input("active", "Execute the Move")]
         [Output("success", "Define if the Move was successful")]
-        public static bool Move(BHoMAdapter source, BHoMAdapter target, IRequest request = null, Dictionary<string, object> pullConfig = null, Dictionary<string, object> pushConfig = null, bool active = false)
+        public static bool Move(BHoMAdapter source, BHoMAdapter target, IRequest request = null, 
+            PullType pullType = PullType.AdapterDefault, ActionConfig pullConfig = null,
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Move.cs]: Is there a reason we are defaulting these variables here and not in the BHoMAdapter itself?

File: BHoM_UI/Components/Adapter/Pull.cs [hunk 2]
```diff
@@ -82,17 +83,45 @@ public override object Run(object[] inputs)
...
-        [Input("request", "Filter on the objects to pull")]
+        [Input("request", "Connect a type of object to Pull objects of that type." +
+            "Connect a Request to pull only objects that satisfy a certain rule.")]
         [Input("config", "Pull config")]
         [Input("active", "Execute the pull")]
-        [Output("objects","Objects pulled")]
-        public static IEnumerable<object> Pull(BHoMAdapter adapter, IRequest request = null, Dictionary<string, object> config = null, bool active = false)
+        [Output("objects", "Objects pulled")]
+        public static IEnumerable<object> Pull(BHoMAdapter adapter, IRequest request = null, 
+            PullType pullType = PullType.AdapterDefault, ActionConfig actionConfig = null,
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Pull.cs]: As above

File: BHoM_UI/Components/Adapter/Pull.cs [hunk 3]
```diff
@@ -82,17 +83,45 @@ public override object Run(object[] inputs)
...
         [Input("config", "Pull config")]
         [Input("active", "Execute the pull")]
-        [Output("objects","Objects pulled")]
-        public static IEnumerable<object> Pull(BHoMAdapter adapter, IRequest request = null, Dictionary<string, object> config = null, bool active = false)
+        [Output("objects", "Objects pulled")]
+        public static IEnumerable<object> Pull(BHoMAdapter adapter, IRequest request = null, 
+            PullType pullType = PullType.AdapterDefault, ActionConfig actionConfig = null, 
+            bool active = false)
         {
+            IRequest actualRequest = null;
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Pull.cs]: As above

File: BHoM_UI/Components/Adapter/Pull.cs [hunk 4]
```diff
@@ -82,17 +83,45 @@ public override object Run(object[] inputs)
...
+            // ---------------------------------------------//
+            // Mandatory Adapter Action set-up              //
+            //----------------------------------------------//
+            // The following are mandatory set-ups to be ALWAYS performed 
+            // before the Adapter Action is called,
+            // whether the Action is overrided at the Toolkit level or not.
+ 
             if (request == null)
-                request = new FilterRequest();
+                actualRequest = new FilterRequest();
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Pull.cs]: Same here. Wouldn't this and all the below, apart from the null check on the `adapter`, be safer be in the BHoMAdapter?

File: BHoM_UI/Components/Adapter/Push.cs [hunk 5]
```diff
@@ -63,15 +64,27 @@ public class PushCaller : MethodCaller
...
-        [Input("config", "Push config")]
+        [Input("pushType", "Push type. Connect the enum PushType for all the alternatives.")]
+        [Input("config", "Push config. Dictionary<string,object> containing any additional data to be used for the Push.")]
         [Input("active", "Execute the push")]
-        [MultiOutput(0, "objects", "Objects that have been pushed (with potentially additional information stored in their CustomData to reflect the push)")]
+        [MultiOutput(0, "objects", "Objects that have been pushed.\nThese objects may be different from the input objects (e.g. their correspondent external software id may be stored in their CustomData).")]
         [MultiOutput(1, "success", "Define if the push was sucessful")]
-        public static Output<List<IObject>, bool> Push(BHoMAdapter adapter, IEnumerable<IObject> objects, string tag = "", Dictionary<string, object> config = null, bool active = false)
+        public static Output<List<object>, bool> Push(BHoMAdapter adapter, IEnumerable<object> objects, string tag = "",
+            PushType pushType = PushType.AdapterDefault, ActionConfig actionConfig = null,
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Push.cs]: Ditto

File: BHoM_UI/Components/Adapter/Remove.cs [hunk 6]
```diff
@@ -46,32 +47,45 @@ public class DeleteCaller : MethodCaller
...
+            }
+
+            // ---------------------------------------------//
+            // Mandatory Adapter Action set-up              //
+            //----------------------------------------------//
+            // The following are mandatory set-ups to be ALWAYS performed 
+            // before the Adapter Action is called,
+            // whether the Action is overrided at the Toolkit level or not.
+
+            if (request == null)
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Remove.cs]: Ditto

File: BHoM_UI/packages.config [hunk 7]
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="MongoDB.Bson" version="2.4.4" targetFramework="net452" />
```
Comment (epignatelli) [inline BHoM_UI/packages.config]: Why is this necessary? Referring the `Serialization_Engine.dll` is not enough?

File: BHoM_UI/BHoM_UI.csproj [hunk 8]
```diff
@@ -63,6 +67,9 @@
       <HintPath>..\..\BHoM_Engine\Build\Library_Engine.dll</HintPath>
       <Private>False</Private>
     </Reference>
+    <Reference Include="MongoDB.Bson, Version=2.4.4.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (epignatelli) [inline BHoM_UI/BHoM_UI.csproj]: Why is this necessary? Referring the `Serialization_Engine.dll` is not enough?

File: BHoM_UI/Templates/Caller.cs [hunk 9]
```diff
@@ -241,22 +244,47 @@ public virtual bool Read(string json)
                 // New serialisation, we stored a CustomObject with SelectedItem, InputParams and OutputParams
                 object backendElement;
                 if (component.CustomData.TryGetValue("SelectedItem", out backendElement))
+                {
+                    if (backendElement == null)
+                    {
+                        BsonDocument bson = Engine.Serialiser.Convert.ToBson(json);
```
Comment (epignatelli) [inline BHoM_UI/Templates/Caller.cs]: I don't think this should be handled here. We have the `Versioning_Engine`, right?

File: BHoM_UI/BHoM_UI.csproj (hunk 8 above)
Comment (alelom) [reply BHoM_UI/BHoM_UI.csproj]: I think that's been added by @adecler's changes.

File: BHoM_UI/Templates/Caller.cs (hunk 9 above)
Comment (alelom) [reply BHoM_UI/Templates/Caller.cs]: I think that's been added by @adecler's changes.

File: BHoM_UI/packages.config (hunk 7 above)
Comment (alelom) [reply BHoM_UI/packages.config]: I think that's been added by @adecler's changes.

File: BHoM_UI/Components/Adapter/Pull.cs (hunk 4 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Pull.cs]: As visible, I didn't add this in for this refactoring. The reason it's there is to assure we have always have a value passed to all Adapter methods.
This cannot be assured in the Adapter because we allow the override of the Adapter Actions.

File: BHoM_UI/Components/Adapter/Move.cs [hunk 10]
```diff
@@ -67,13 +68,30 @@ public class MoveCaller : MethodCaller
...
+                return false;
+            }
+
+            // ---------------------------------------------//
+            // Mandatory Adapter Action set-up              //
+            //----------------------------------------------//
+            // The following are mandatory set-ups to be ALWAYS performed 
+            // before the Adapter Action is called,
+            // whether the Action is overrided at the Toolkit level or not.
+
```
Comment (alelom) [inline BHoM_UI/Components/Adapter/Move.cs]: This is the reason why we want to have some initialisation in the UI itself rather than later in the adapter.
We allow for Action overloading and we want to assure that certain initialisation are done no matter the overload.

File: BHoM_UI/Components/Adapter/Move.cs (hunk 1 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Move.cs]: As per our chat and as further explained [here](https://github.com/BHoM/BHoM_UI/pull/167/files/40e23834c3eb21462d165709731bbffa938f3866#r354452867)
there are certain initialisation that must be assured before the Adapter methods (aka Actions) are called, because we allow for them to be overloaded.

File: BHoM_UI/Components/Adapter/Pull.cs (hunk 2 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Pull.cs]: See my comment here https://github.com/BHoM/BHoM_UI/pull/167#discussion_r354453473

File: BHoM_UI/Components/Adapter/Pull.cs (hunk 3 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Pull.cs]: See my comment here #167 (comment)

File: BHoM_UI/Components/Adapter/Pull.cs (hunk 4 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Pull.cs]: See my comment here #167 (comment)

File: BHoM_UI/Components/Adapter/Push.cs (hunk 5 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Push.cs]: See my comment here #167 (comment)

File: BHoM_UI/Components/Adapter/Remove.cs (hunk 6 above)
Comment (alelom) [reply BHoM_UI/Components/Adapter/Remove.cs]: See my comment here #167 (comment)

**Comment by adecler** [comment]
@epignatelli , some of the versioning needs to happen on the level of the UI, hence the link to Mongo.Bson.

**Comment by alelom** [comment]
> `+ $exception {"Unable to cast object of type 'MongoDB.Bson.BsonNull' to type 'MongoDB.Bson.BsonDocument'."} System.InvalidCastException`
> 
> This PR is messing with the CustomObjects,
> Try creating a customObject with properties, then copy and paste it or close and re-open grasshopper.
> Afterwards the assigned properties disappear.
> image

Thanks for this @kThorsager. 
I had a look, I think this could be linked to one of the changes done by @adecler, either on:
- [`Caller.cs`](https://github.com/BHoM/BHoM_UI/pull/167/files#diff-57fe9b87f51a6223525a669c505b65e6)
- [`CreateObject.cs`](https://github.com/BHoM/BHoM_UI/pull/167/files#diff-59b7ada80c65d7f337cc8717427ab4aa)
@adecler could you please help me bottoming this out?

**Comment by adecler** [comment]
Thanks @kThorsager , good (and critical!) catch 😅 .

I have fixed it in an additional commit. This issue was that I had forgotten to handle the case where `SelectedItem` is null before serialising it (which is always the case for `CreateCustom`). 

Should all be fine now but let me know if you still find problems.

**Comment by alelom** [comment]
Thanks for this @adecler!

File: BHoM_UI/Components/Adapter/Pull.cs [hunk 11]
```diff
@@ -80,19 +81,43 @@ public override object Run(object[] inputs)
...
         {
+            // ---------------------------------------------//
+            // Mandatory Adapter Action set-up              //
+            //----------------------------------------------//
+            // The following are mandatory set-ups to be ALWAYS performed 
+            // before the Adapter Action is called,
+            // whether the Action is overrided at the Toolkit level or not.
+
+            // If unset, set the actionConfig to a new ActionConfig.
+            actionConfig = actionConfig == null ? new ActionConfig() : actionConfig;
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Pull.cs]: All of this happens because we allow users to override the `Pull` method, in this case. If we prevent that, we can avoid spreading a lot of checks in other projects. This is explained in the Adapter pr https://github.com/BHoM/BHoM_Adapter/pull/164

File: BHoM_UI/Components/Adapter/Pull.cs [hunk 12]
```diff
@@ -80,19 +81,43 @@ public override object Run(object[] inputs)
...
+            // Always assure there is a Request. Allow to input a Type to generate a FilterRequest.
+            IRequest actualRequest = null;
+
             if (request == null)
-                request = new FilterRequest();
+                actualRequest = new FilterRequest();
+            else if (request is Type)
+                actualRequest = BH.Engine.Data.Create.FilterRequest((Type)request, "");
+            else if (typeof(IRequest).IsAssignableFrom(request.GetType()))
+                actualRequest = request as IRequest;
```
Comment (epignatelli) [inline BHoM_UI/Components/Adapter/Pull.cs]: Just to reinforce, this code is here because we allow the override of the `Pull` method and. because of that, we are not sure the `base.Pull` will be called - thus, we cannot enforce this logic.

The more I read the code, the more I think that the `Pull` method should not be `virtual`, and anything the user wants to do, can be done in a `BeforePull` and `AfterPull` methods.

File: BHoM_UI/Components/Adapter/Pull.cs (hunk 11 above)
Comment (adecler) [reply BHoM_UI/Components/Adapter/Pull.cs]: We need to be able to override the `Pull` in some of the lower level adapters

Comment (alelom) [reply BHoM_UI/Components/Adapter/Pull.cs]: I agree with both - a mechanism like the one suggested by @epignatelli in https://github.com/BHoM/BHoM_Adapter/pull/164 would still be advisable in my mind, so we can limit overrides to only the _strictly neededed_ cases.

Comment (adecler) [reply BHoM_UI/Components/Adapter/Pull.cs]: > so we can limit overrides to only the strictly neededed cases.

Please no! There is nothing more painful when you are hacking with someone else code than to realise that the stuff you want to change is not overridable.

Comment (alelom) [reply BHoM_UI/Components/Adapter/Pull.cs]: @adecler that's why I said "limit" and didn't say "disallow completely"!
Could you please confirm you would oppose a strategy where we have:
- overridable "Before[SomeActionName]" and "After[SomeActionName]" (where ActionName is Push/Pull/etc), that run before and after each individual adapter action, and
- overridable Actions (in which the override is permitted, but discouraged unless strictly needed)?

Comment (adecler) [reply BHoM_UI/Components/Adapter/Pull.cs]: I would. Anything that doesn't allow to completely override the adapter actions is a step back for me.  `Before` and `After` overridable methods still means that you have no say on code being executed in the middle. There are better ways to guide new contributors without limiting advanced developers than restricting access to code.
